languages/java/oso/src/main/java/com/osohq/oso/Oso.java

```diff
@@ -91,7 +105,173 @@ public boolean isAllowed(Object actor, Object action, Object resource)
         .collect(Collectors.toCollection(HashSet::new));
   }
 
-  public static void main(String[] args) throws Exceptions.OsoException, IOException {
-    new Oso().repl(args);
+  /**
+   * Ensure that `actor` is allowed to perform `action` on `resource`.
+   *
+   * <p>If the action is permitted with an `allow` rule in the policy, then this method returns
+   * `None`. If the action is not permitted by the policy, this method will raise an error.
+   *
+   * <p>The error raised by this method depends on whether the actor can perform the `"read"` action
+   * on the resource. If they cannot read the resource, then a `NotFound` error is raised.
+   * Otherwise, a `ForbiddenError` is raised.
+   *
+   * @param actor The actor performing the request.
+   * @param action The action the actor is attempting to perform.
+   * @param resource The resource being accessed.
+   * @param checkRead If set to `false`, a `ForbiddenError` is always thrown on authorization
+   *     failures, regardless of whether the actor can read the resource. Default is `true`.
+   * @throws Exceptions.OsoException
+   */
+  public void authorize(Object actor, Object action, Object resource, boolean checkRead)
+      throws Exceptions.OsoException {
+    boolean authorized = queryRuleOnce("allow", actor, action, resource);
+    if (authorized) {
+      return;
+    }
+
+    // Authorization failure. Determine whether to throw a NotFoundException or
+    // a ForbiddenException.
+    boolean isNotFound = false;
+    if (action == readAction) {
+      isNotFound = true;
+    } else if (checkRead) {
+      boolean canRead = queryRuleOnce("allow", actor, readAction, resource);
+      if (!canRead) {
+        isNotFound = true;
+      }
+    }
+    throw isNotFound ? new Exceptions.NotFoundException() : new Exceptions.ForbiddenException();
+  }
+
+  public void authorize(Object actor, Object action, Object resource)
+      throws Exceptions.OsoException {
+    authorize(actor, action, resource, true);
+  }
+
+  /**
+   * Ensure that `actor` is allowed to send `request` to the server.
+   *
+   * <p>Checks the `allow_request` rule of a policy.
+   *
+   * <p>If the request is permitted with an `allow_request` rule in the policy, then this method
+   * returns nothing. Otherwise, this method raises a `ForbiddenError`.
+   *
+   * @param actor The actor performing the request.
+   * @param request An object representing the request that was sent by the actor.
+   * @throws Exceptions.OsoException
+   */
+  public void authorizeRequest(Object actor, Object request) throws Exceptions.OsoException {
```

Comment:
This can still throw an `OsoException` I think -- e.g. if your policy throws an error unrelated to enforcement (same goes for the other methods you called out here)

Reply:
Oh duh